File: problems/transx_problem.py
from PyGMO.problem import base as base_problem
from PyKEP.core import epoch, DAY2SEC, MU_SUN, lambert_problem, propagate_lagrangian, fb_prop, fb_vel, AU
from PyKEP.planet import jpl_lp,spice
from math import sqrt, pi, cos, sin, acos, atan2, log
from scipy.linalg import norm
from numpy import *

excludes = []

class transx_problem(base_problem):
    """
    This class is a PyGMO (http://esa.github.io/pygmo/) problem representing a Multiple Gravity Assist
    trajectory allowing one only impulsive Deep Space Manouvre between each leg.

    - Izzo, Dario. "Global optimization and space pruning for spacecraft trajectory design." Spacecraft Trajectory Optimization 1 (2010): 178-200.

    The decision vector is::

      [t0, T] + [u, v, Vinf, eta1, a1] + [beta, rp/rV, eta2, a2] + ...

    ... in the units: [mjd2000, days] + [nd,nd,km/s,nd,years,nd] + [rad,nd,nd,nd] + ....

    Each leg time-of-flight can be decoded as follows, T_n = T log(alpha_n) / \sum_i(log(alpha_i)).

    .. note::

       The resulting problem is box-bounded (unconstrained). The resulting trajectory is time-bounded.
    """

    def exclude(self, x):
        excludes.append(x)

    def __init__(self, seq=[jpl_lp('earth'), jpl_lp('venus'), jpl_lp('earth')], dim = 1, obj_dim  = 1, avoid = []):

        # Sanity checks ...... all planets need to have the same
        # mu_central_body
        if ([r.mu_central_body for r in seq].count(seq[0].mu_central_body) != len(seq)):
            raise ValueError('All planets in the sequence need to have exactly the same mu_central_body')
        self.n_legs = len(seq) - 1
        
        # We then define all planets in the sequence  and the common central
        # body gravity as data members
        self.seq = seq
        self.common_mu = seq[0].mu_central_body
        if self.common_mu == 0:
          self.common_mu = 1e10
        self.avoid = []
        # First we call the constructor for the base PyGMO problem
        # As our problem is n dimensional, box-bounded (may be multi-objective), we write
        # (dim, integer dim, number of obj, number of con, number of inequality con, tolerance on con violation)
        super(transx_problem, self).__init__(dim, 0, obj_dim, 0, 0, 0)


    def print_time_info(self, planets, times):
        # print out the results

        encounters = [] 
        transfers = []
        for i in range(len(planets) - 1):
          src = planets[i]
          dst = planets[i + 1]
          transfers.append(['Transfer time from ' + src.name + ' to ' + dst.name + ':', str(round(times[i + 1].mjd - times[i].mjd, 2)), " days"])

        for i in range(len(planets)):
          x = planets[i]
          encounters.append("Date of " + x.name + ' encounter: ' + str(times[i]))

        for enc in encounters:
          print(enc)


        max_l = max(map(lambda x: len(x[0]), transfers))

        transfers = map(lambda x: x[0].ljust(max_l) + x[1] + x[2], transfers)
        for tran in transfers:
          print(tran)
            
        print "Total mission duration:             ", round(times[-1].mjd - times[0].mjd, 2), " days"
        print
        print
       
        
    def print_escape(self, ref, V_ref, R_ref, deltaV, eject_T):
        dVTransX = self.velocity_to_transx(V_ref, R_ref, deltaV)
        dv = norm(deltaV)

        print "TransX escape plan - " + ref.name + " escape"
        print "--------------------------------------"
        print("MJD:                 %10.4f "    % round(eject_T, 4))
        print("Prograde:            %10.3f m/s" % round(dVTransX[0], 3))
        print("Outward:             %10.3f m/s" % round(dVTransX[1], 3))
        print("Plane:               %10.3f m/s" % round(dVTransX[2], 3))
        print("Hyp. excess velocity:%10.3f m/s" % round(dv, 3))
        print("Earth escape burn:   %10.3f m/s" % round(self.burn_cost(ref, deltaV), 3))
        print
        print 

    def print_dsm(self, V_ref, R_ref, deltaV, v, dsm_T):
        dVTransX = self.velocity_to_transx(V_ref, R_ref, deltaV)
        dv = norm(deltaV)

        print "Deep Space Burn"
        print "--------------------------------------"
        print("MJD:                 %10.4f "    % round(dsm_T, 4))
        print("Prograde:            %10.3f m/s" % round(dVTransX[0], 3))
        print("Outward:             %10.3f m/s" % round(dVTransX[1], 3))
        print("Plane:               %10.3f m/s" % round(dVTransX[2], 3))
        print("Hyp. excess velocity:%10.3f m/s" % round(norm(v), 3))
        print("DSM burn:            %10.3f m/s" % round(dv, 3))
        print
        print 
        

    def burn_cost(self, ref, exc, circ = True):
        mu = ref.mu_self
        r = ref.radius + 200000 
        dv = []
        return sqrt(dot(exc, exc) + 2 * mu / r) - sqrt((1 if circ else 2) * mu / r)
        
    def print_flyby(self, ref, V_ref, R_ref, Vin, Vout, enc_T):
        VoutTransX = self.velocity_to_transx(V_ref, R_ref, Vout)

        vx = VoutTransX[0]
        vy = VoutTransX[1]
        vz = VoutTransX[2]

        print ref.name + " encounter"
        print "--------------------------------------"
        print("MJD:                 %10.4f "    % round(enc_T, 4))
        print("Approach velocity:   %10.3f m/s" % round(sqrt(dot(Vin,Vin)), 3))
        print("Departure velocity:  %10.3f m/s" % round(sqrt(dot(Vout,Vout)), 3))  
        print("Outward angle:       %10.3f deg" % round(180*atan2(vy, vx)/pi, 3))
        print("Inclination:         %10.3f deg" % round(180*atan2(vz, sqrt(vx * vx + vy * vy))/pi, 3))
        ta  = acos(dot(Vin, Vout)/sqrt(dot(Vin,Vin))/sqrt(dot(Vout,Vout)))
        print("Turning angle:       %10.3f deg" % round(ta*180/pi, 3))
        alt = (ref.mu_self / dot(Vin,Vin)*(1/sin(ta/2)-1) - ref.radius)/1000
        print("Periapsis altitude:  %10.3f km " % round(alt, 3))
        print("dV needed:           %10.3f m/s" % round(fb_vel(Vin, Vout, ref), 3))
        print
        print 
 
    def print_arrival(self, ref, V_exc, enc_T):

        print ref.name + " arrival"
        print "--------------------------------------"
        print("MJD:                 %10.4f    " % round(enc_T, 4))
        print("Hyp. excess velocity:%10.3f m/s" % round(norm(V_exc), 3))
        print("Orbit insertion burn %10.3f m/s" % round(self.burn_cost(ref, V_exc), 3))
        print
        print

    def velocity_to_transx(self, v_ref, v_rad, v):
        fward = v_ref / linalg.norm(v_ref)
        plane = cross(v_ref, v_rad); plane = plane / linalg.norm(plane)
        oward = cross(plane, fward)

        vx = dot(fward, v)
        vy = dot(oward, v)
        vz = dot(plane, v)

        return [vx, vy, vz]

    def calc_objective(self, x, should_print = False):
   
        if self.f_dimension == 1:
            return (0,)
        else:
            return (0, 0)

    # Objective function
    def _objfun_impl(self, x):
        if len(self.avoid) > 0:
          d = max([(exp(norm(x) / (norm([a - b for a, b in zip(x, y)]))) - 1) for y in self.avoid])
          return d + self.calc_objective(x)


        return self.calc_objective(x)

    def pretty(self, x):
        # Plot of the trajectory
        self.calc_objective(x, True)

File: src/problems/transx_problem.cpp
#include <string>
#include <vector>
#include <numeric>
#include <cmath>
#include <iostream>
#include <boost/math/constants/constants.hpp>
#include <boost/array.hpp>

#include <keplerian_toolbox/core_functions/array3D_operations.h>
#include <keplerian_toolbox/core_functions/fb_vel.h>

#include "transx_problem.h"

namespace pagmo { namespace problem {

transx_problem::transx_problem(const std::vector<kep_toolbox::planet::planet_ptr> seq,
    const double dep_altitude, const double arr_altitude, const bool circularize,
    const int dim, const int obj_dim) : base(dim, 0, obj_dim, 0, 0, 0.0), m_n_legs(seq.size() - 1), m_arr_altitude(arr_altitude), m_dep_altitude(dep_altitude), m_circularize(circularize) 
{
  std::vector<double> mus(seq.size());
  for (std::vector<kep_toolbox::planet::planet_ptr>::size_type i = 0; i < seq.size(); ++i) {
    mus[i] = seq[i]->get_mu_central_body();
  }
  if ((unsigned int)std::count(mus.begin(), mus.end(), mus[0]) != mus.size()) {
    pagmo_throw(value_error, "The planets do not all have the same mu_central_body");
  }

  for (std::vector<kep_toolbox::planet::planet_ptr>::size_type i = 0; i < seq.size(); ++i) {
    m_seq.push_back(seq[i]->clone());
  }

  m_common_mu = seq[0]->get_mu_central_body();
  m_max_deltaV = 20000.0;
}

transx_problem::transx_problem(const transx_problem &p) : base(p.get_dimension(), 0, p.get_f_dimension(), 0, 0, 0, 0), m_seq(), m_common_mu(p.m_common_mu), m_n_legs(p.m_n_legs), m_circularize(p.m_circularize), m_arr_altitude(p.m_arr_altitude), m_dep_altitude(p.m_dep_altitude) {
  for (std::vector<kep_toolbox::planet::planet_ptr>::size_type i = 0; i < p.m_seq.size(); ++i) {
    m_seq.push_back(p.m_seq[i]->clone());
  }

  set_bounds(p.get_lb(),p.get_ub());
} 

base_ptr transx_problem::clone() const {
  return base_ptr(new transx_problem(*this));
}

std::ostream& operator<<(std::ostream& ss, const TransXTimes &tim) {
  ss << std::fixed;

  std::vector<std::string> encounters;
  std::vector<std::pair<std::string, std::string> > transfers;
  for (int i = 0; i < tim.planets().size() - 1; ++i) {
    ss << "Transfer time from " << tim.planets().Get(i) << " to " << tim.planets().Get(i+1) << ":";
    ss << (tim.times().Get(i + 1) - tim.times().Get(i)) << std::setprecision(2) << " days" << std::endl;
  }

  for (int i = 0; i < tim.planets().size(); ++i) {
    ss << "Date of " << tim.planets().Get(i) << " encounter: ";
    ss << tim.times().Get(i) << std::endl;
  }

  ss << "Total mission duration: " << (tim.times().Get(tim.times().size() - 1) - tim.times().Get(0)) << std::setprecision(2) << " days" << std::endl << std::endl << std::endl;

  return ss;
}

void transx_problem::transx_time_info(TransXTimes *times, std::vector<kep_toolbox::planet::planet_ptr> planets, std::vector<kep_toolbox::epoch> time_list) const {
  for (int i = 0; i < planets.size(); ++i) {
    times->add_planets(planets[i]->get_name());
    times->add_times(time_list[i].mjd());
  }
}

std::ostream& operator<<(std::ostream& ss, const TransXEscape &esc) {
  ss << std::fixed;
  ss << "Escape - " << esc.planet() << std::endl;
  ss << "--------------------------------------" << std::endl;
  ss << "MJD:                     " << esc.mjd() << std::setprecision(4) << std::endl;
  ss << "Prograde:                " << esc.prograde() << std::setprecision(3) << std::endl;
  ss << "Outward:                 " << esc.outward() << std::setprecision(3) << std::endl;
  ss << "Plane:                   " << esc.plane() << std::setprecision(3) << std::endl;
  ss << "Hyp. excess velocity:    " << esc.vinf() << std::setprecision(3) <<  " m/s" << std::endl;
  ss << "Earth escape burn:       " << esc.burn() << std::setprecision(3) << std::endl;
  ss << std::endl << std::endl;

  return ss;
};

void transx_problem::transx_escape(TransXEscape *escape, kep_toolbox::planet::planet_ptr ref, kep_toolbox::array3D V_ref, kep_toolbox::array3D R_ref, kep_toolbox::array3D deltaV, double eject_T) const {

  kep_toolbox::array3D dVTransX = velocity_to_transx(V_ref, R_ref, deltaV);
  double dv = kep_toolbox::norm(deltaV);

  escape->set_planet(ref->get_name());
  escape->set_mjd(eject_T);
  
  escape->set_prograde(dVTransX[0]);
  escape->set_outward(dVTransX[1]);
  escape->set_plane(dVTransX[2]);

  escape->set_vinf(dv);

  escape->set_burn(burn_cost(ref, deltaV, false, true));
}

std::ostream& operator<<(std::ostream& ss, const TransXDSM &dsm) {

  ss << std::fixed;
  ss << "Deep Space Maneuver - " << std::endl;
  ss << "--------------------------------------" << std::endl;
  ss << "MJD:                     " << dsm.mjd() << std::setprecision(4) << std::endl;
  ss << "Prograde:                " << dsm.prograde() << std::setprecision(3) << std::endl;
  ss << "Outward:                 " << dsm.outward() << std::setprecision(3) << std::endl;
  ss << "Plane:                   " << dsm.plane() << std::setprecision(3) << std::endl;
  ss << "Hyp. excess velocity:    " << dsm.vinf() << std::setprecision(3) <<  "m/s" << std::endl;
  ss << "DSM burn:                " << dsm.burn() << std::setprecision(3) << std::endl;
  
  ss << std::endl << std::endl;

  return ss;
}

void transx_problem::transx_dsm(TransXDSM * dsm, kep_toolbox::array3D V_ref, kep_toolbox::array3D R_ref, kep_toolbox::array3D deltaV, kep_toolbox::array3D v, double dsm_T) const {
  kep_toolbox::array3D dVTransX = velocity_to_transx(V_ref, R_ref, deltaV);
  double dv = kep_toolbox::norm(deltaV);

  dsm->set_mjd(dsm_T);

  dsm->set_prograde(dVTransX[0]);
  dsm->set_outward(dVTransX[1]);
  dsm->set_plane(dVTransX[2]);

  dsm->set_vinf(kep_toolbox::norm(v));

  dsm->set_burn(dv);
}

std::ostream& operator<<(std::ostream& ss, const TransXFlyBy &f) {
  ss << std::fixed;
  ss << f.planet() << " encounter" << std::endl;
  ss << "--------------------------------------" << std::endl;
  ss << "MJD:                   " << f.mjd() << std::setprecision(4) << std::endl;
  ss << "Approach velocity:     " << f.approach_vel() << std::setprecision(3) << std::endl;
  ss << "Departure velocity:    " << f.departure_vel() << std::setprecision(3) << std::endl;  

  ss << "Outward angle:         " << f.outward_angle() << std::endl;
  ss << "Inclination:           " << f.inclination() << std::setprecision(3) << " deg" << std::endl;
  ss << "Turning angle:         " << f.turning_angle() << std::setprecision(3) << " deg" << std::endl;
  ss << "Periapsis altitude:    " << f.periapsis_altitude() << std::setprecision(3) << " km" << std::endl;
  ss << "dV needed:             " << f.burn() << std::setprecision(3) << " m/s" << std::endl;
  ss << std::endl << std::endl;

  return ss;
}

void transx_problem::transx_flyby(TransXFlyBy *flyby, kep_toolbox::planet::planet_ptr ref, kep_toolbox::array3D V_ref, kep_toolbox::array3D R_ref, kep_toolbox::array3D Vin, kep_toolbox::array3D Vout, double enc_T) const {
  kep_toolbox::array3D VoutTransX = velocity_to_transx(V_ref, R_ref, Vout);
  double vx = VoutTransX[0];
  double vy = VoutTransX[1];
  double vz = VoutTransX[2];

  flyby->set_planet( ref->get_name() );
  flyby->set_mjd( enc_T );

  flyby->set_approach_vel( sqrt(kep_toolbox::dot(Vin, Vin)) );
  flyby->set_departure_vel( sqrt(kep_toolbox::dot(Vout, Vout)) );
  
  flyby->set_outward_angle( 180 * atan2(vy, vx) / M_PI );
  flyby->set_inclination( 180 * atan2(vz, sqrt(vx * vx + vy * vy)) / M_PI );

  double ta  = acos(kep_toolbox::dot(Vin, Vout)/sqrt(kep_toolbox::dot(Vin,Vin))/sqrt(kep_toolbox::dot(Vout,Vout)));
  flyby->set_turning_angle( ta * 180.0 / M_PI );

  double alt = (ref->get_mu_self() / kep_toolbox::dot(Vin,Vin)*(1/sin(ta/2)-1) - ref->get_radius())/1000;
  flyby->set_periapsis_altitude( alt );

  double enc_vel = 0.0;
  kep_toolbox::fb_vel(enc_vel, Vin, Vout, *ref);
  flyby->set_burn(enc_vel);
}

std::ostream& operator<<(std::ostream& ss, const TransXArrival &arriv) {
  ss << std::fixed;
  ss << arriv.planet() << " arrival" << std::endl;
  ss << "--------------------------------------" << std::endl;
  ss << "MJD:                     " << arriv.mjd() << std::setprecision(4) << "   " << std::endl;
  ss << "Hyp. excess velocity:    " << arriv.vinf() << std::setprecision(3) <<  "m/s" << std::endl;
  ss << "Orbit insertion burn:    " << arriv.burn() << std::setprecision(3) << " m/s" << std::endl;
  ss << std::endl << std::endl;

  return ss;
}

void transx_problem::transx_arrival(TransXArrival *arrival, kep_toolbox::planet::planet_ptr ref, kep_toolbox::array3D V_exc, double enc_T) const {

  arrival->set_planet(ref->get_name());
  arrival->set_mjd(enc_T);
  arrival->set_vinf(kep_toolbox::norm(V_exc));
  arrival->set_burn(burn_cost(ref, V_exc, true, m_circularize));
}

std::ostream& operator<<(std::ostream& ss, const TransXSolution &sol)
{
  ss << sol.times();

  ss << sol.escape();

  int i = 0; 
  int j = 0;
  auto dsms = sol.dsms();
  auto flybyes = sol.flybyes();
  while (i < dsms.size() || j < flybyes.size()) {
    if (i < dsms.size()) {
      ss << dsms.Get(i++);
    }
    if (j < flybyes.size()) {
      ss << flybyes.Get(j++);
    }
  }

  ss << sol.arrival();

  ss << "Total delta-V:           " << sol.fuel_cost() << std::setprecision(3) << " m/s" << std::endl;

  return ss;
}

kep_toolbox::array3D transx_problem::velocity_to_transx(kep_toolbox::array3D v_ref, kep_toolbox::array3D v_rad, kep_toolbox::array3D v) const {
  
  kep_toolbox::array3D fward;
  kep_toolbox::array3D plane;
  kep_toolbox::array3D oward;

  kep_toolbox::vers(fward, v_ref);
  kep_toolbox::cross(plane, v_ref, v_rad); 
  kep_toolbox::vers(plane, plane);
  kep_toolbox::cross(oward, plane, fward);

  kep_toolbox::array3D res;
  res[0] = kep_toolbox::dot(fward, v);
  res[1] = kep_toolbox::dot(oward, v);
  res[2] = kep_toolbox::dot(plane, v);

  return res;
}

double transx_problem::burn_cost(kep_toolbox::planet::planet_ptr ref, const kep_toolbox::array3D &exc, bool arr, bool circ) const {
  double mu = ref->get_mu_self();
  double r = ref->get_radius() + (arr ? m_arr_altitude : m_dep_altitude) * 1000;
  double norm_sqr = kep_toolbox::dot(exc, exc);
  return sqrt(norm_sqr + 2 * mu / r) - sqrt((circ ? 1 : 2) * mu / r);
}

TransXSolution transx_problem::calc_objective(fitness_vector &f, const decision_vector &x, bool should_print) const {
  TransXSolution solution;
  f[0] = 0;
  if (get_f_dimension() == 2) {
    f[1] = 0;
  }
  return solution;
}

void transx_problem::objfun_impl(fitness_vector &f, const decision_vector &x) const {
  calc_objective(f, x);
}

TransXSolution transx_problem::get_solution(const decision_vector &x, bool extended_output) const {
  fitness_vector f(get_f_dimension());
  auto solution = calc_objective(f, x, true);
  for (int i =0; i < x.size(); ++i) {
    solution.add_x(x[i]);
  }
  solution.set_problem(get_name());
  return solution;
}

std::string transx_problem::get_name() const {
  return "TransX base problem";
}

std::string transx_problem::human_readable_extra() const {
  return "TransX base problem - Do not use, subclass";
}

}} // namespaces

BOOST_CLASS_EXPORT_IMPLEMENT(pagmo::problem::transx_problem)
